### PR TITLE
expose sparse mv/mm algo selection

### PR DESCRIPTION
**Pull request by Roger-luo**
this is a simple change to expose the algorithm parameter to mv/mm level

edit: I'm not sure if we should check if the selected algo match the input, e.g some algo only work on CSR with col/row-major

### CODE REVIEW

**Comment by Roger-luo** [comment]
I thought the current tests cover this already tho? Or we should test on other algorithms?

**Comment by maleadt** [comment]
> I thought the current tests cover this already tho?

How? Didn't you introduce the `algo` keyword here?

File: test/cusparse/generic.jl [hunk 1]
```diff
@@ -0,0 +1,47 @@
+using Test
```
Comment (maleadt) [inline test/cusparse/generic.jl]: Don't import all these. If you want to run a single test, so `julia -L test/setup.jl test/cusparse/generic.jl`.

File: test/cusparse/generic.jl [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+using Random
+using CUDA
+using CUDA.CUSPARSE, SparseArrays
+
+Random.seed!(123)
```
Comment (maleadt) [inline test/cusparse/generic.jl]: Seed is already applied in setup.jl

File: lib/cusparse/generic.jl [hunk 3]
```diff
@@ -82,8 +82,10 @@ Base.unsafe_convert(::Type{cusparseSpMatDescr_t}, desc::CuSparseMatrixDescriptor
 
 ## API functions
 
-function mv!(transa::SparseChar, alpha::Number, A::Union{CuSparseMatrixBSR{T},CuSparseMatrixCSR{T}},
-             X::DenseCuVector{T}, beta::Number, Y::DenseCuVector{T}, index::SparseChar) where {T}
+function mv!(transa::SparseChar, alpha::Number,
```
Comment (maleadt) [inline lib/cusparse/generic.jl]: Better keep the current formatting, like the rest of the file.

File: lib/cusparse/generic.jl [hunk 4]
```diff
@@ -102,18 +104,21 @@ function mv!(transa::SparseChar, alpha::Number, A::Union{CuSparseMatrixBSR{T},Cu
...
         cusparseSpMV(handle(), transa, Ref{compute_type}(α), descA, descX, Ref{compute_type}(β),
-                     descY, compute_type, CUSPARSE_MV_ALG_DEFAULT, buffer)
+                     descY, compute_type, algo, buffer)
     end
     Y
 end
 
-function mv!(transa::SparseChar, alpha::Number, A::CuSparseMatrixCSC{T}, X::DenseCuVector{T},
-             beta::Number, Y::DenseCuVector{T}, index::SparseChar) where {T}
+function mv!(transa::SparseChar, alpha::Number,
```
Comment (maleadt) [inline lib/cusparse/generic.jl]: same.

**Review by maleadt — CHANGES_REQUESTED** [review]

**Comment by Roger-luo** [comment]
hmm I'm a bit confused why the test is failing on CUDA 11.1, I've filter the tests to not run below CUDA 11.2, but somehow it still get executed on those versions, what would be right way of doing this?

**Comment by maleadt** [comment]
`CUDA.version()` returns the driver version. You need to check against `CUSPARSE.version()` (which requires you to use the library version, so look up what the appropriate version is corresponding with CUDA toolkit 11.2 -- it's in the release notes, or you can check yourself by setting `JULIA_CUDA_VERSION=11.2`).

File: test/cusparse/generic.jl [hunk 5]
```diff
@@ -1,7 +1,6 @@
-using CUDA
 using CUDA.CUSPARSE, SparseArrays
 
-if CUDA.version() >= v"11.2" # lower CUDA version doesn't support these algorithms
+if CUSPARSE.version() >= v"11.2" # lower CUDA version doesn't support these algorithms
```
Comment (maleadt) [inline test/cusparse/generic.jl]: This is wrong, and that's what I meant with having to check the version.

```
$ JULIA_CUDA_VERSION=11.2 jl -e 'using CUDA; @show CUSPARSE.version()'                                                                                       
  Downloaded artifact: CUDA
CUSPARSE.version() = v"11.4.1"
```

**Comment by maleadt** [comment]
Please rebase on master to not require Windows to pass.

**Review by maleadt — APPROVED** [review]

**Comment by maleadt** [comment]
LGTM, thanks!

**Pull request merged**